### ISSUE TITLE
ngMock.$httpBackend.verifyNoOutstandingRequest don't work as expected

### ISSUE THREAD

**Issue by gimmi** [angular/angular.js#5453]
Consider the following failing test:

```
describe('verifyNoOutstandingRequest does not work', function () {

    it('should flush call', inject(function ($http, $httpBackend) {
        $httpBackend.when('POST', '/api/feedback', {}).respond('done');

        var result = '';
        $http.post('/api/feedback', {}).then(function (ret) {
            result = ret.data;
        });

        expect(result).toBe('');
        $httpBackend.flush();
        expect(result).toBe('done');
        $httpBackend.verifyNoOutstandingRequest();
    }));

    it('should detect outstanding request', inject(function ($http, $httpBackend) {
        var result = '';
        $http.post('/api/feedback', {}).then(function (ret) {
            result = ret.data;
        });

        expect(function () {
            $httpBackend.verifyNoOutstandingRequest();
        }).toThrow('Unflushed requests: 1');
    }));
});
```

The test `should detect outstanding request` doesn't work. I'm using `angular-1.2.5` and `jasmine-2.0.0`

**Comment by bengro** [comment]
We (@bengro and @tomwhoiscontrary) are seeing what we believe to be the same bug. Here are some test cases for handling of unstubbed requests which we believe should all pass:

```javascript
describe('$httpBackend', function () {
    var $httpBackend, $http, $rootScope;

    beforeEach(inject(function (_$httpBackend_, _$http_, _$rootScope_) {
        $httpBackend = _$httpBackend_;
        $http = _$http_;
        $rootScope = _$rootScope_;
    }));

    it('should fail to flush if there are unexpected requests', function () {
        $http.get('http://example.org');

        expect($httpBackend.flush).toThrow();
    });

    it('should fail to digest if there are unexpected requests', function () {
        $http.get('http://example.org');

        expect($rootScope.$digest).toThrow();
    });

    it('should fail verify that there are no outstanding requests if there are unexpected requests', function () {
        $http.get('http://example.org');

        expect($httpBackend.verifyNoOutstandingRequest).toThrow(); // fails
    });

    it('should fail to verify that there are no outstanding expectations if there are unexpected requests', function () {
        $http.get('http://example.org');

        expect($httpBackend.verifyNoOutstandingExpectation).toThrow();
    });
});
```

The test for `verifyNoOutstandingRequest` fails. The documentation for `verifyNoOutstandingRequest` states that it "verifies that there are no outstanding requests that need to be flushed". However, if we make a request, do not flush, and then call `verifyNoOutstandingRequest`, it returns silently.

Have we misunderstood what `verifyNoOutstandingRequest` is for? Are we misusing it? Or is there a bug?

**Comment by kentcdodds** [comment]
I'm seeing this as well. I'm using angular version `1.3.15`.

**Comment by ManFromEast** [comment]
I got the same problem, in fact I want to test the below program:
item.like = function (item) {

      if (item.isLiked) {
        api.likes.unlike($.param({collect_id: item.id})).$promise.then(
          function () {
            if (item.isLiked) {
              if (angular.isDefined(SiteCatalyst.pageName)) {
                delete SiteCatalyst.pageName;  
              }      
              item.isLiked = false;
              item.likesCount--;
            }
          }
        );
      }
      else {
        api.likes.like($.param({collect_id: item.id})).$promise.then(
          function () {
            if (!item.isLiked) {
              item.isLiked = true;
              item.likesCount++;
            }
          }
        );
      }
    };


Testing Code:
  describe("ItemService", function() {
    var $rootScope, ItemService, httpBackend, item;

    beforeEach(inject(function (_$rootScope_, _ItemService_, _$httpBackend_) {
      $rootScope = _$rootScope_;
      httpBackend = _$httpBackend_;
      ItemService = _ItemService_;
      item = {id: 1700000022202202, likesCount: 5};
    }));

    afterEach(function() {
      httpBackend.verifyNoOutstandingRequest();
      httpBackend.verifyNoOutstandingExpectation();
    });

    describe("#like", function() {
      describe("when user already like the item", function() {
        beforeEach(function() {
          item.isLiked = true;
        });
        it("should call DELETE /api/like/collect API", function() {
          httpBackend.expect("POST", "/api/like/collect", {collect_id: 1700000022202202})
            .respond({"status":"success","code":200,"data":"ok","meta":{"hits":1}});
          ItemService.like(item);
          $rootScope.$digest();
          httpBackend.flush();
          
          expect(item.isLiked).toBeFalsy();
        });
      });

      describe("when user haven't like the item", function() {
        beforeEach(function() {
          item.isLiked = false;
        });
        it("should call POST /api/like/collect API", function() {
          httpBackend.expect("POST", "/api/like/collect", {collect_id: 1})
            .respond({});

          ItemService.like(item);
          httpBackend.flush();
          expect(item.isLiked).toBeTruthy();
        });
      });
    });
  });

but after the program go into like function, it can't reach the api, it popup the below error:
IE 11.0.0 (Windows 7 0.0.0) [Service] ItemService #like when user already like the item should call POST /api/like/collect API FA
        Error: Unexpected request: POST https://domain/api/like/collect
        No more request expected
           at $httpBackend (c:/xxx/js/angular-mocks/angular-mocks.js:1178:5)
           at s (c:/xx/js/angular/angular.min.js:74:109)
           at Anonymous function (c:/xx//js/angular/angular.min.js:71:422)
           at J (c:/xx//js/angular/angular.min.js:101:76)
           at J (c:/xx//js/angular/angular.min.js:101:76)
           at Anonymous function (c:/xx//js/angular/angular.min.js:102:249)
           at h.prototype.$eval (c:/xx//js/angular/angular.min.js:113:21)
           at h.prototype.$digest (c:/xx//js/angular/angular.min.js:110:109)
           at Anonymous function (c:/xx/spec/services/itemSpec.js:35:11)

**Comment by Narretz** [comment]
I assume this was fixed by https://github.com/angular/angular.js/commit/267ee9c892b0eb40908700ee2435793f8c6c1c84

**Comment by waldemar-schwarz** [comment]
Hi guys,

Please excuse me if this is not the right place to ask. My team is working with Angular 1.4.x, and we would need this fix to be available on that branch of Angular as well.

Is it possible to create a PR for this fix on 1.4.x branch?

Kind regards,
Waldemar

**Comment by gkalpak** [comment]
Sorry, @waldemar-schwarz, the 1.4.x branch is no longer supported.
And even if it were, the fix (267ee9c89) involved a breaking change, so we couldn't introduce it to that branch anyway.

**Comment by waldemar-schwarz** [comment]
Understood. Thank you @gkalpak for checking & explaining.
